google-services-plugin/src/main/kotlin/com/google/gms/googleservices/GoogleServicesTask.kt

```diff
@@ -0,0 +1,366 @@
+        val resValues: MutableMap<String, String?> =
+            TreeMap() // TreeMap to preserve order with previous plugin versions
+        handleProjectNumberAndProjectId(rootObject, resValues)
+        handleFirebaseUrl(rootObject, resValues)
+        val clientObject = getClientForPackageName(rootObject)
+        if (clientObject != null) {
+            handleAnalytics(clientObject, resValues)
+            handleMapsService(clientObject, resValues)
+            handleGoogleApiKey(clientObject, resValues)
+            handleGoogleAppId(clientObject, resValues)
+            handleWebClientId(clientObject, resValues)
+        } else {
+            throw GradleException("No matching client found for package name '${applicationId.get()}' in ${quickstartFile.path}")
+        }
+
+        // write the values file.
+        val values = File(intermediateDir, "values")
+        if (!values.exists() && !values.mkdirs()) {
+            throw GradleException("Failed to create folder: $values")
+        }
+        File(values, "values.xml").writeText(
+            getValuesContent(resValues), Charsets.UTF_8
+        )
+    }
+
+    @Throws(IOException::class)
+    private fun handleFirebaseUrl(rootObject: JsonObject, resValues: MutableMap<String, String?>) {
+        val projectInfo = rootObject.getAsJsonObject("project_info")
+            ?: throw GradleException("Missing project_info object")
+        val firebaseUrl = projectInfo.getAsJsonPrimitive("firebase_url")
+        if (firebaseUrl != null) {
+            resValues["firebase_database_url"] = firebaseUrl.asString
+        }
+    }
+
+    /**
+     * Handle project_info/project_number for @string/gcm_defaultSenderId, and fill the res map with
+     * the read value.
+     *
+     * @param rootObject the root Json object.
+     * @throws IOException
+     */
+    @Throws(IOException::class)
+    private fun handleProjectNumberAndProjectId(
+        rootObject: JsonObject, resValues: MutableMap<String, String?>
+    ) {
+        val projectInfo = rootObject.getAsJsonObject("project_info")
+            ?: throw GradleException("Missing project_info object")
+        val projectNumber = projectInfo.getAsJsonPrimitive("project_number")
+            ?: throw GradleException("Missing project_info/project_number object")
+        resValues["gcm_defaultSenderId"] = projectNumber.asString
+        val projectId = projectInfo.getAsJsonPrimitive("project_id")
+            ?: throw GradleException("Missing project_info/project_id object")
+        resValues["project_id"] = projectId.asString
+        val bucketName = projectInfo.getAsJsonPrimitive("storage_bucket")
+        if (bucketName != null) {
+            resValues["google_storage_bucket"] = bucketName.asString
+        }
+    }
+
+    private fun handleWebClientId(
+        clientObject: JsonObject, resValues: MutableMap<String, String?>
+    ) {
+        val array = clientObject.getAsJsonArray("oauth_client")
```

Comment:
Instead of iterating through an array of the object, I'd just deserialize it into a data class. Could use `kotlinx.serialization` to do this as well.

Reply:
sure, although I'd like to limit this to a mostly 1:1 translation of the Groovy code to Kotlin, focusing only on required AGP API changes, not on plugin business logic changes. If I find time later I'd be happy to optimize / make the plugin code more maintainable (or someone else can take over and do it), but I'd like to keep it out of scope of this PR if that's OK.